src/chains/ethereum/ethereum/src/forking/handlers/provider-handler.ts

```diff
@@ -0,0 +1,77 @@
+import { EthereumInternalOptions } from "@ganache/ethereum-options";
+import { BaseHandler } from "./base-handler";
+import { Handler } from "../types";
+import { AbortSignal } from "abort-controller";
+
+import { JsonRpcError, JsonRpcResponse } from "@ganache/utils";
+
+type JsonRpc = JsonRpcResponse | JsonRpcError;
+export class ProviderHandler extends BaseHandler implements Handler {
+  private _request: (
+    method: string,
+    params: unknown[]
+  ) => Promise<{
+    response: { result: any } | { error: { message: string; code: number } };
+    raw: string;
+  }>;
+  constructor(options: EthereumInternalOptions, abortSignal: AbortSignal) {
+    super(options, abortSignal);
+    const provider = options.fork.provider;
+
+    if (typeof provider.request === "function") {
+      this._request = async (method: string, params: unknown[]) => {
+        try {
+          const result = await provider.request({ method, params });
+          const response = { result };
+          return { response, raw: JSON.stringify(response) };
+        } catch (error: any) {
+          // if this doesn't appear to be a JSON-RPC "coded" error it might be
+          // a network level error, or something else we don't know how to
+          // handle. Throw so stack traces are preserved.
+          if (typeof error.code !== "number") throw error;
+          return {
+            response: { error },
+            raw: null
+          };
+        }
+      };
+    } else if (typeof (provider as any).send === "function") {
+      // TODO: remove support for legacy providers' legacy `.send`
+      console.warn(
+        "WARNING: Ganache forking only supports EIP-1193-compliant providers. Legacy support for send is currently enabled, but will be removed in a future version _without_ a breaking change. To remove this warning, switch to an EIP-1193 provider. This error is probably caused by an old version of Web3's HttpProvider (or ganache < v7)"
+      );
+      this._request = async (method: string, params: unknown[]) => {
+        return await new Promise((resolve, reject) => {
+          const request = {
+            id: this.id++,
+            jsonrpc: "2.0",
+            method,
+            params
+          };
+          (provider as any).send(request, (err: Error, response: JsonRpc) => {
+            if (err) return void reject(err);
+            resolve({
+              response,
+              raw: JSON.stringify(response)
+            });
+          });
+        });
+      };
+    } else {
+      throw new Error("Forking `provider` must be EIP-1193 compatible");
+    }
+  }
+  public async request<T>(method: string, params: unknown[]) {
+    // format params via JSON stringification because the params might
+    // be Quantity or Data, which aren't valid as `params` themselves,
+    // but when JSON stringified they are
+    const strParams = JSON.stringify(params);
+
+    return await this.queueRequest<T>(`${method}:${strParams}`, () =>
```

Comment:
Above in `http-handler`, the `key` for `queueRequest` is `JSON.stringify({ method, params })`, which won't be equal to the `key` for this request, `'${method}:${strParams}'`. Does that matter?